src/cpp/shared/adventure/adv.cpp

```diff
@@ -341,6 +374,248 @@ void advManager::ComputerMonsterInteract(mapCell *cell, hero *computer, int *a1)
+    }
+    
+    int xOffset = 0;
+    int xOffset2 = (192 / numCreaturesSecondRow - 32) / 2 + 22;
+    int xOffset3 = 192 / numCreaturesSecondRow;
+
+    int widgetIdx = 0;
+    int garrisonIdx = 0;
+    for(int i = 0; i < numCreaturesSecondRow; ++i) {
+      if(garrisonCreatures == 5) {
+        if(i)
+          xOffset = -12;
+        else
+          xOffset = 12;
+      }
+      while(town->garrison.creatureTypes[garrisonIdx] == -1)
+        ++garrisonIdx;
+
+      int imgIdx = town->garrison.creatureTypes[garrisonIdx];
+
+      IconEntry *iconEntry = GetIconEntry(res, imgIdx);
+
+      int v10 = xOffset + xOffset2 + xOffset3 * widgetIdx - iconEntry->offsetX;
+      int x = (32 - iconEntry->width) / 2 + v10 + 1;
+      int y = (yOffset - iconEntry->offsetY) - iconEntry->height + 30;
+
+      iconWidgetArray[widgetIdx] = new iconWidget(x, y, 32, 32, "mons32.icn", imgIdx, 0, -1, 16, 1);
+      if(!iconWidgetArray[widgetIdx])
+        MemError();           
+
+      x = xOffset + xOffset2 + xOffset3 * widgetIdx - 14;
+      y = yOffset + 32;
+      char *quantityStr = this->GetQuantityString(thievesGuildsLevel, town, garrisonIdx);
+      textWidgetArray[widgetIdx] = new textWidget(x, y, 60, 12, quantityStr, "smalfont.fnt", 1, -1, 512, 1);
+      if(!textWidgetArray[widgetIdx])
+        MemError();
+
+      window->AddWidget((widget *)iconWidgetArray[widgetIdx], -1);
+      window->AddWidget((widget *)textWidgetArray[widgetIdx++], -1);
+      ++garrisonIdx;
+    }
+
+    if(numCreaturesFirstRow) {
+      xOffset3 = 192 / numCreaturesFirstRow;
+      int v31 = (192 / numCreaturesFirstRow - 32) / 2 + 22;
+      int yOffset2 = yOffset + 44;
+      for(int i = numCreaturesSecondRow; numCreaturesFirstRow + numCreaturesSecondRow > i; ++i) {
+        while(town->garrison.creatureTypes[garrisonIdx] == -1)
+          ++garrisonIdx;
+        int imgIdx = town->garrison.creatureTypes[garrisonIdx];
+
+        int v14 = yOffset2 - GetIconEntry(res, imgIdx)->offsetY;
+        int y = v14 - GetIconEntry(res, imgIdx)->height + 30;
+        int v16 = v31 + xOffset3 * (widgetIdx - numCreaturesSecondRow) - GetIconEntry(res, imgIdx)->offsetX;
+        IconEntry *v17 = GetIconEntry(res, imgIdx);
+        int x = (32 - v17->width) / 2 + v16 + 1;
+        iconWidgetArray[widgetIdx] = new iconWidget(x, y, 32, 32, "mons32.icn", imgIdx, 0, -1, 16, 1);
+
+        if(!iconWidgetArray[widgetIdx])
+          MemError();       
+
+        x = v31 + xOffset3 * (widgetIdx - numCreaturesSecondRow) - 14;
+        y = yOffset2 + 32;
+        char *quantityStr = this->GetQuantityString(thievesGuildsLevel, town, garrisonIdx);
+        textWidgetArray[widgetIdx] = new textWidget(x, y, 60, 12, quantityStr, "smalfont.fnt", 1, -1, 512, 1);
+
+        if(!textWidgetArray[widgetIdx])
+          MemError();
+        window->AddWidget((widget *)iconWidgetArray[widgetIdx], -1);
+        window->AddWidget((widget *)textWidgetArray[widgetIdx++], -1);
+        ++garrisonIdx;
+      }
+    }
```

Comment:
Please take a pass over the variable names in this function.

Reply:
This one is insanity. I don't think I have enough vocabulary to describe these variables